### DISCUSSION TITLE
Rule Severity or additional information

### DISCUSSION THREAD

**Discussion by rfcdejong** [MarimerLLC/csla#2502] [Ideas]
Currently business rules only know "Error, Warning, Informational or Success" as Rule Severity.
We have business rules that result into Warnings, but there must be a weight defined, sometimes a warning means a property should have a value for information. other times a warning means the property should had a value for calculations to even continue. 

I don't see a way to provide extra information (other than hacking this into the Description).

Can I override RuleResults into the Context with an extra property for "DetailedSeverity" or introduce a custom RuleSeverity?

**Comment by hurcane** [comment]
When you say "a warning means a property should have a value for information", what exactly does that mean? From my outsider perspective, that sounds like how one would describe an information severity. Why do you consider that different from the existing information level? Perhaps some specific examples would help with determining if there might be a way to solve your use case with the existing levels.

**Comment by rfcdejong** [comment]
We use information for information icon next to fields to show what the field means or just tips for it.

We use errors for fields that are mandatory or have wrong input to prevent saving the business object, for example
  - an date of birth in the future
  - an date of birth more than xxx years in the past
  
We use warnings for fields that:
  - are minimally required to fill and so essential for our program to function
  - are required being information about something, like a date of birth should have a value to calculate other fields (age, pension start date, etc)
  - are required to have a value for finalizing a data request towards an external provider
  
We worked around this using conditionals based on a boolean, but that is just on a few fields and cumbersome.
Either errors or warnings should have some sort of weight or label to add to a field.

**Reply by hurcane** [reply]
If you had a weight attribute available, how would your code behave differently for the different weights? Do you have a specific example of different warnings that would get different severities?

**Reply by rfcdejong** [reply]
a picture might exlain 

![image](https://user-images.githubusercontent.com/2665653/135108538-8ce17e27-4ada-47ce-9eab-0b5eadd8dc25.png)

**Reply by hurcane** [reply]
How do these label check boxes relate to a business rule that would have a weight?

**Reply by rfcdejong** [reply]
They relatie 1:1 always, but each label is related to a 'warning'.
Ofcourse there are several ways to rome, we could just add each single business rule or property to a list and then configure labels. However then we lose dynamic results based on other property values.

It just would had been great if we just could add a label to a warning inside the busines rule. But I don't see a way overriding or adding anything to the RuleResult.

**Comment by Chicagoan2016** [comment]
I had a similar use case, I used "rule chaining", if memory serves, with gateway rule (creating business objects book in Using Csla 4 series)

**Comment by hurcane** [comment]
Would sub-classing CommonBusinessRule (e.g. WeightedBusinessRule) and adding a property for the weight meet your needs in a cleaner way? The code that needs to evaluate the weight can cast the business rule to the WeightedBusinessRule class when needed.

**Reply by rfcdejong** [reply]
for static outcome we could sub-class it, but then dynamic results based on other properties won't be possible

**Reply by hurcane** [reply]
Rules are not static. They have a primary property and the option for additional input properties. If any of those properties changes, the rule is re-evaluated. If your rule needs "dynamic results based on other properties", then those properties should be an input to the rule so it can be triggered automatically when those properties change.

**Reply by rfcdejong** [reply]
an extra output property foreach primairy property, and an additional weight property foreach property. lots of coding to do, but it will work. just hoped for something easier.
